### PR TITLE
Fixed the bug in redirect exposed in #1049

### PR DESCRIPTION
**Pull request by ShraddheyaS**
#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/reply.js [hunk 1]
```diff
@@ -122,6 +122,7 @@ Reply.prototype.headers = function (headers) {
 
 Reply.prototype.code = function (code) {
   this.res.statusCode = code
+  this._headers['statusCode'] = code
```
Comment (mcollina) [inline lib/reply.js]: Can you please use a `_hasStatusCode = true` property instead?

File: lib/reply.js [hunk 2]
```diff
@@ -147,6 +148,10 @@ Reply.prototype.redirect = function (code, url) {
     code = 302
   }
 
+  if (this._headers.hasOwnProperty('statusCode')) {
+    code = this._headers['statusCode']
+  }
```
Comment (mcollina) [inline lib/reply.js]: I would avoid this block, and avoid calling `code(code)`.

File: lib/reply.js [hunk 3]
```diff
@@ -122,6 +122,7 @@ Reply.prototype.headers = function (headers) {
 
 Reply.prototype.code = function (code) {
   this.res.statusCode = code
+  this._headers['_hasStatusCode'] = true
```
Comment (delvedor) [inline lib/reply.js]: You can directly use `this._hasStatusCode`, without changing the headers object :)

Comment (ShraddheyaS) [reply lib/reply.js]: Okay, but how do I add the `_hasStatusCode` property to `this`? Or let me guess, it is initialized automatically?

Comment (delvedor) [reply lib/reply.js]: You must initialize it here: https://github.com/fastify/fastify/blob/36960965710756148cd86bc040d2c6f25ee4d04d/lib/reply.js#L24-L33 The default value should be `false`.

**Comment by ShraddheyaS** [comment]
While writing the unit tests, does the case like `res.code(302).redirect(200, '/')` need to be handled?

**Comment by mcollina** [comment]
Yes I think so

File: lib/reply.js [hunk 4]
```diff
@@ -142,12 +144,22 @@ Reply.prototype.type = function (type) {
...
-  this.header('location', url).code(code).send()
+  if (this._hasStatusCode) {
+    finalCode = this.res.statusCode
+  }
+
+  if (typeof code === 'number') {
+    finalCode = code
+  }
+
+  this.header('location', finalUrl).code(finalCode).send()
```
Comment (delvedor) [inline lib/reply.js]: I think you can refactor the function in this way:
```js
Reply.prototype.redirect = function (code, url) {
  if (typeof code === 'string') {
    url = code
    code = this._hasStatusCode ? this.res.statusCode : 302
  }

  this.header('location', url).code(code).send()
}
```
Does it work?

**Pull request merged**